File: Data Structures & Algorithms/duplicate-integer/submission-3.java
class Solution {
    public boolean hasDuplicate(int[] nums) {
        Set<Integer> hashSet = new HashSet<Integer>();

        for(int num: nums){
            if(hashSet.contains(num))
                return true;
            hashSet.add(num);
        }

        return false;
    }
}

File: Data Structures & Algorithms/maximum-product-subarray/submission-3.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        maxP, minP = 1, 1
        res = nums[0]
        for num in nums:
            if num == 0:
                maxP, minP = 1, 1
                res = max(res, 0)
                continue
            t = maxP
            maxP = max(num, num*maxP, num*minP)
            minP = min(num, num*t, num*minP)
            res = max(res, maxP, minP)
        
        return res